pkg/controller/common.go

```diff
@@ -597,335 +600,805 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
+						if err != nil {
+							continue
+						}
 
-			updatedTolerations := []v1.Toleration{}
-			for _, podToleration := range podTolerations {
-				if existingToleration, ok := podTolerationMap[podToleration.Key]; ok {
-					if podToleration.MatchToleration(existingToleration) {
-						updatedTolerations = append(updatedTolerations, podToleration)
+						nodeChecker := set{}
+						for _, attachment := range attachments {
+							if i == 0 {
+								klog.V(5).Infof("Adding node (%s) to the replica node set: replica mounts of volume (%s) found on node", attachment.Spec.NodeName, volume)
+								replicaNodes.add(attachment.Spec.NodeName)
+							} else {
+								nodeChecker.add(attachment.Spec.NodeName)
+							}
+						}
+						if i > 0 {
+							// take an intersection of the current pod list's replica nodes with those of preceding pod lists.
+							for node := range replicaNodes {
+								if !nodeChecker.has(node) {
+									klog.V(5).Infof("Removing node (%s) from the replica node set: replica mounts of volume (%s) cannot be found on node", node, volume)
+									replicaNodes.remove(node)
+								}
+							}
+						}
+					}
+				}
+
+				qualifiedNodes := set{}
+				for node := range nodesWithSelectedPods {
+					var nodeInstance v1.Node
+					if err = p.azr.getClient().Get(ctx, types.NamespacedName{Name: node.(string)}, &nodeInstance); err != nil {
+						klog.Errorf("failed to get node (%s)", node.(string))
+						continue
+					}
+					if value, exists := nodeInstance.Labels[podAffinity.TopologyKey]; exists {
+						// get list of node candidates
+						nodeRequirement, err := azureutils.CreateLabelRequirements(podAffinity.TopologyKey, selection.Equals, value)
+						if err != nil {
+							continue
+						}
+						nodeSelector := labels.NewSelector().Add(*nodeRequirement)
+
+						nodeList := &v1.NodeList{}
+						if err = p.azr.getClient().List(ctx, nodeList, &client.ListOptions{LabelSelector: nodeSelector}); err != nil {
+							klog.Errorf("failed to list nodes that satisfy labels selector (%v): %v", nodeSelector, err)
+							continue
+						}
+
+						for _, node := range nodeList.Items {
+							qualifiedNodes.add(node.Name)
+						}
+					}
+				}
+
+				// remove any candidate node which does not satisfy the pod affinity
+				for candidateNode := range candidateNodes {
+					if !qualifiedNodes.has(candidateNode) {
+						klog.Infof("Removing node (%s) from candidate nodes: node does not satisfy inter-pod-affinity (%v)", candidateNode, podAffinity)
+						candidateNodes.remove(candidateNode)
 					}
 				}
 			}
-			podTolerations = updatedTolerations
 		}
+	}
 
-		nodeSelector := labels.SelectorFromSet(labels.Set(podObj.Spec.NodeSelector))
-		requirements, selectable := nodeSelector.Requirements()
-		if selectable {
-			podNodeSelector = podNodeSelector.Add(requirements...)
+	filteredNodes := []v1.Node{}
+	for replicaNode := range replicaNodes {
```

Comment:
In this loop, we could check if nodeMap contains replicaNode and if it does, we can add it to the candidateNode list. Then we won't need the loop on line 752 to verify if the nodeMap contains candidateNodes, as the candidateNodes list and the nodeMap started out with the same set of nodes. 

We can then return candidateNodes at the end.